python/pyarrow/scalar.pxi

```diff
@@ -838,6 +843,67 @@ cdef class UnionScalar(Scalar):
         return sp.type_code
 
 
+cdef class ExtensionScalar(Scalar):
+    """
+    Concrete class for Extension scalars.
+    """
+
+    @property
+    def value(self):
+        """
+        Return storage value as a scalar.
+        """
+        cdef CExtensionScalar* sp = <CExtensionScalar*> self.wrapped.get()
+        return Scalar.wrap(sp.value) if sp.is_valid else None
+
+    def as_py(self):
+        """
+        Return this scalar as a Python object.
+        """
+        # XXX should there be a hook to wrap the result in a custom class?
```

Comment:
Similarly as we have `__arrow_ext_class__` for using an ExtensionArray subclass for the array object, we could indeed have a way to specify a custom scalar wrapper (but that's also fine to leave as a future enhancement I think)

Reply:
Feel free to open a JIRA, as you probably know the desired behaviour better than me :-)

Reply:
Opened https://issues.apache.org/jira/browse/ARROW-13612